### PR TITLE
thread local: allow to destroy slot on worker thread

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: thread local: allow to destroy slot on worker thread
Additional Description:

The thread local slot should be allocated and destroyed on the main thread. But it's a little complex to ensure this point if we use the slot in the RDS because the route may be destroyed on the worker.

This PR do a minor update to the slot implementation to ensure we can destroy a slot on the worker thread. After this PR, we needn't the clean up logic like following any more:

```
  ~LuaClusterSpecifierConfig() {
    // The design of the TLS system does not allow TLS state to be modified in worker threads.
    // However, when the route configuration is dynamically updated via RDS, the old
    // LuaClusterSpecifierConfig object may be destructed in a random worker thread. Therefore, to
    // ensure thread safety, ownership of per_lua_code_setup_ptr_ must be transferred to the main
    // thread and destroyed when the LuaClusterSpecifierConfig object is not destructed in the main
    // thread.
    if (per_lua_code_setup_ptr_ && !main_thread_dispatcher_.isThreadSafe()) {
      auto shared_ptr_wrapper =
          std::make_shared<PerLuaCodeSetupPtr>(std::move(per_lua_code_setup_ptr_));
      main_thread_dispatcher_.post([shared_ptr_wrapper] { shared_ptr_wrapper->reset(); });
    }
  }
```

If the slot is destructed when the server is running as usual, the PR only make sure the `removeSlot()` to be called at main thread and won't break anything. There would no any lifetime issue.

If the slot is destructed when the server is shutting, the state would be more complex, but it should also be fine. Because the shutting will clear all slots anyway. 

The shutting process is as following:
1. Main dispatcher exits.
2. Set the shutdown flag of TLS.
3. Stop all workers and will destroy all slots, pending callbacks on workers. All data, states will be destroyed.
4. Destroy all slots on main thread.
5. Destroy all pending callbacks on workers.
6. Destroy whole server.

Risk Level: mid. Simple update to very core code.
Testing: unit.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Review by mattklein123 — CHANGES_REQUESTED** [review]
At face value this seems fine but it's a pretty risky change. I would probably runtime guard it.

/wait

**Comment by wbpcode** [comment]
@mattklein123 I think of a new idea to make it simpler and safer. Rather than change the destructor of `SlotImpl` and affect all the use cases, we can just provide a similar wrapper like the TypeSlot to do the thread-safely clean up.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Wouldn't we always want this behavior though? It seems like we should add the runtime guard and have this be the default and then cleanup all the callsites that work around this? wdyt?

/wait

**Comment by wbpcode** [comment]
> Wouldn't we always want this behavior though? It seems like we should add the runtime guard and have this be the default and then cleanup all the callsites that work around this? wdyt?

Ideally, yeah. It would be great to make the new logic a default behavior. And it should be safe in logic.

But another fact is that almost all current slot needn't the new logic except the slots related to route. Changing the destructor still make me afraid because it's so core and important structure, even I added a runtime guard.

What I want actually is just a simpler way to use slot in the route. So, a new wrapper like `ThreadSafeTypeSlot` which will handle the clean up automatically is also fine for me.

It up to you or the community. I implemented both. I just reverted the refactor and now the latest code is the version with new destructor and runtime guard.

And the new wrapper way is simpler which is only a new template class like following:

```
template <class T> class ThreadSafeTypedSlot : public TypedSlot<T> {
public:
  /**
   * Helper method to create a unique_ptr for a thread safe typed slot. This helper
   * reduces some verbose parameterization at call-sites.
   *
   * @param allocator factory to allocate untyped Slot objects.
   * @param main_thread_dispatcher the dispatcher for the main thread.
   * @return a TypedSlotPtr<T> (the type is defined below).
   */
  static std::unique_ptr<ThreadSafeTypedSlot>
  makeUnique(SlotAllocator& allocator, Event::Dispatcher& main_thread_dispatcher) {
    return std::make_unique<ThreadSafeTypedSlot>(allocator, main_thread_dispatcher);
  }

  explicit ThreadSafeTypedSlot(SlotAllocator& allocator, Event::Dispatcher& main_thread_dispatcher)
      : TypedSlot<T>(allocator), main_thread_dispatcher_(main_thread_dispatcher) {}

  /**
   * Destructor that will post the destruction of the slot to the main thread if the slot is not
   * being destroyed on the main thread.
   */
  ~ThreadSafeTypedSlot() {
    if (!main_thread_dispatcher_.isThreadSafe()) {
      main_thread_dispatcher_.post([slot = std::move(this->slot_)] {});
    }
    this->slot_.reset();
  }

private:
  Event::Dispatcher& main_thread_dispatcher_;
};

template <class T = ThreadLocalObject>
using ThreadSafeTypedSlotPtr = std::unique_ptr<ThreadSafeTypedSlot<T>>;
```

File: changelogs/current.yaml [hunk 1]
```diff
@@ -16,6 +16,11 @@ behavior_changes:
     Changes the default value of ``envoy.reloadable_features.http2_use_oghttp2`` to ``false``. This changes the codec used for HTTP/2
     requests and responses. A number of users have reported issues with oghttp2 including issue 32611 and issue 32401 This behavior
     can be reverted by setting the feature to ``true``.
+- area: thread_local
+  change: |
+    Changes the behavior of the ``SlotImpl`` class destructor. By this change, now the destructor could be called at any thread.
+    This behavior can be reverted by setting the runtime flag ``envoy.reloadable_features.allow_slot_destroy_on_worker_threads``
+    to false.
```
Comment (mattklein123) [inline changelogs/current.yaml:23]:
```suggestion
    Changes the behavior of the ``SlotImpl`` class destructor. With this change the destructor can be called on any thread.
    This behavior can be reverted by setting the runtime flag ``envoy.reloadable_features.allow_slot_destroy_on_worker_threads``
    to false.
```

**Review by mattklein123 — APPROVED** [review]
LGTM. I personally prefer the real fix and then we can do cleanups elsewhere as needed.

**Comment by wbpcode** [comment]
/wait it's near to cut 1.30 major release, let's pend this up and go head after the realeasing.

**Comment by wbpcode** [comment]
gentle ping for a new review or LGTM. Thanks cc @mattklein123

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**